subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/resolver/KotlinBuildScriptModelRepository.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.resolver
+
+import org.gradle.kotlin.dsl.concurrent.ResurrectingThread
+import org.gradle.kotlin.dsl.tooling.models.KotlinBuildScriptModel
+
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.suspendCoroutine
+
+
+private
+typealias AsyncModelRequest = Pair<KotlinBuildScriptModelRequest, Continuation<KotlinBuildScriptModel?>>
+
+
+internal
+open class KotlinBuildScriptModelRepository {
+
+    open suspend fun scriptModelFor(request: KotlinBuildScriptModelRequest): KotlinBuildScriptModel? =
+        suspendCoroutine { k ->
+            accept(request, k)
+            processor.poke()
+        }
+
+    protected
+    open fun accept(request: KotlinBuildScriptModelRequest, k: Continuation<KotlinBuildScriptModel?>) {
+        q.push(request to k)
+    }
+
+    protected
+    open fun fetch(request: KotlinBuildScriptModelRequest): KotlinBuildScriptModel =
+        fetchKotlinBuildScriptModelFor(request)
+
+    private
+    val q = ConcurrentGroupingQueue<AsyncModelRequest> {
+        first.scriptFile == it.first.scriptFile && first.projectDir == it.first.projectDir
+    }
+
+    private
+    val processor = ResurrectingThread("Kotlin DSL Resolver") {
+        while (true) {
+            val group = q.nextGroup()
+            if (group.isEmpty()) {
+                break
+            }
+            process(group)
+        }
+    }
+
+    private
+    fun process(group: List<AsyncModelRequest>) {
+        val (mostRecentRequest, k) = group.first()
+        val result = runCatching { fetch(mostRecentRequest) }
+        resumeAll(supersededRequestsFrom(group), Result.success(null))
+        resume(k, result)
```

Comment:
❓shouldn't this be resumed without `ignoringErrors {}` ?

Reply:
No, we don't want an exception in the client to cause the resolver thread to quit.

Reply:
Right, I get it we don't bother about the cancelled requests. But I think that if this statement fails there's a real problem and we should log that, did I miss something?

Reply:
The logging should have happened on the client before propagating the exception. And if we need to diagnose some related error in the future, the `printStackTrace` should be enough (the output should appear in the IntelliJ log).

Notice this is the same behaviour as before.

Reply:
Got it now, thanks. SGTM